Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeOverlayRegionLayers.mm

```diff
@@ -0,0 +1,163 @@
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "RemoteLayerTreeOverlayRegionLayers.h"
+
+#if ENABLE(OVERLAY_REGIONS_IN_EVENT_REGION)
+
+#import "PlatformCALayerRemote.h"
+#import "RemoteLayerTreeHost.h"
+#import "WKScrollView.h"
+#import "WKWebViewIOS.h"
+#import <QuartzCore/QuartzCore.h>
+#import <WebCore/IntRectHash.h>
+#import <objc/runtime.h>
+#import <pal/spi/cocoa/QuartzCoreSPI.h>
+
+@interface UIView (UIViewInternal_RemoteLayerTreeOverlayRegionLayers)
+- (UIViewController *)_viewControllerForAncestor;
+@end
+
+static inline NSData* dataFromRect(CGRect rect)
+{
+    return [NSData dataWithBytes:(void *)&rect length:sizeof(CGRect)];
+}
+
+@interface WKScrollView (RemoteLayerTreeOverlayRegionLayersAdditions)
+@property (nonatomic, readonly) NSArray<NSData *> *overlayRegions;
+@property (nonatomic, readonly) NSMapTable<UIView *, NSData *> *internalOverlayRegions;
+@end
+
+@implementation WKScrollView (RemoteLayerTreeOverlayRegionLayersAdditions)
+- (NSArray<NSData *> *)overlayRegions
+{
+    return NSAllMapTableValues(self.internalOverlayRegions);
+}
+
+- (NSMapTable<UIView *, NSData *> *)internalOverlayRegions
+{
+    static void* internalOverlayRegionsKey = &internalOverlayRegionsKey;
+
+    NSMapTable *mapTable = objc_getAssociatedObject(self, internalOverlayRegionsKey);
+    if (mapTable)
+        return mapTable;
+
+    mapTable = [NSMapTable weakToStrongObjectsMapTable];
+    objc_setAssociatedObject(self, internalOverlayRegionsKey, mapTable, OBJC_ASSOCIATION_RETAIN_NONATOMIC);
+
+    return mapTable;
+}
+@end
+
+namespace WebKit {
+using namespace WebCore;
+
+static bool isInScrollableRegion(UIView *view, const RemoteLayerTreeHost& layerTreeHost)
+{
+    extern NSString *fixedScrollableNodeLayerKey;
+
+    for ( ; view; view = [view superview]) {
+        CALayer *layer = view.layer;
+        if (RemoteLayerTreeNode *node = RemoteLayerTreeNode::forCALayer(layer)) {
+            if (auto scrollContainerID = node->actingScrollContainerID()) {
+                RemoteLayerTreeNode *scollingContainerNode = layerTreeHost.nodeForID(scrollContainerID);
+                CALayer *scrollLayer = scollingContainerNode ? scollingContainerNode->uiView().layer : nil;
+                if (auto value = (NSNumber*)[scrollLayer valueForKey:fixedScrollableNodeLayerKey])
+                    return value.intValue ? false : true;
+            }
+        }
+
+        if (auto value = (NSNumber*)[layer valueForKey:fixedScrollableNodeLayerKey])
+            return value.intValue ? false : true;
+
+        for (CALayer *sublayer in layer.sublayers) {
```

Comment:
I'm not sure what's going on here. The outer loop walks up the view hierarchy, but there's a sublayer traversal here?

Reply:
I'll investigate this and confirm. I found the ScrollingStateFixedNode's layer to be a sublayer, maybe the view hierarchy transversal is not even necessary.

Reply:
I debugged more and I find the ScrollingStateFixedNode's layer on the node's second superview's layer's second sublayer, i.e.,:
```
[(CALayer*)(((UIView*)[[nodeView superview] superview]).layer.sublayers[2]) valueForKey:@"WKFixedScrollableNodeLayerKey"]	NSConstantIntegerNumber *	0x283c88088	0x0000000283c88088
NSNumber	NSNumber	
_encoding	char *const	"i"	0x000000028307d7cc
_value	long long	1
```
where `nodeView` is the one in `updateDebugOverlayRectForView`